tools/wheel/build-wheels

```diff
@@ -0,0 +1,329 @@
+
+    # Generate extra arguments to specify what stage to build.
+    if stage is not None:
+        extra = ['--target', stage]
+    else:
+        extra = []
+
+    # Run the build.
+    print('[-] Build', tag, extra + args)
+    docker('build', '--tag', tag, *extra, *args, script_dir)
+
+    return tag
+
+
+def target_args(target):
+    """
+    Returns the docker build arguments for the specified platform target.
+    """
+    platform_name = target.platform_name
+    platform_version = target.platform_version
+    python_version = target.python_version
+    return [
+        '--build-arg', f'PYTHON={python_version[0]}.{python_version[1:]}',
+        '--build-arg', f'PLATFORM={platform_name}:{platform_version}',
+    ]
+
+
+def build_image(target, identifier, options):
+    """
+    Runs the build for a target and (optionally) extract the wheel.
+    """
+    args = [
+        '--ssh', 'default',
+        '--build-arg', f'DRAKE_VERSION={options.version}',
+    ] + target_args(target)
+    if not options.keep_containers:
+        args.append('--force-rm')
+
+    # Build the image.
+    if options.tag_stages:
+        # Inspect Dockerfile, find stages, and build them.
+        for line in open(os.path.join(script_dir, 'Dockerfile')):
+            if line.startswith('FROM'):
+                stage = line.strip().split()[-1]
+                tag = build_stage(target, args, tag_prefix=stage, stage=stage)
+    else:
+        tag = build_stage(target, args, tag_prefix=identifier)
+        images_to_remove.append(tag)
+
+    # Extract the wheel (if requested).
+    if options.extract:
+        print('[-] Extracting wheel(s) from', tag)
+
+        command = 'tar -cf - /wheel/wheelhouse/*.whl'
+        extractor = docker(
+            'run', '--rm', tag, 'bash', '-c', command, pipe=True)
+        subprocess.check_call(
+            ['tar', '--strip-components=2', '-xf', '-'],
+            stdin=extractor.stdout, cwd=options.output_dir)
+
+        extractor_result = extractor.wait()
+        if extractor_result:
+            raise subprocess.CalledProcessError(extractor_result,
+                                                extractor.args, None, None)
+
+
+def test_wheel(target, options):
+    """
+    Runs the test script for the wheel matching the specified target.
+    """
+    vm = f'cp{target.python_version}'
+    wheel = f'drake-{options.version}-{vm}-{vm}m-manylinux_2_27_x86_64.whl'
+
+    platform = target.platform_alias
+    container = f'{tag_base}:test-{platform}-py{target.python_version}'
+    test_dir = os.path.join(script_dir, 'test')
+
+    docker('build', '-t', container, *target_args(target), test_dir)
+
+    docker('run', '--rm', '-t',
+           '-v' f'{test_dir}:/test',
+           '-v' f'{options.output_dir}:/wheel',
+           container, '/test/test-wheel.sh', f'/wheel/{wheel}')
+
+
+def check_version(version):
+    """Returns True iff the given version string matches PEP 440."""
```

Comment:
> Can we document somehow what set the giant regex is supposed to match? I understand it may difficult or error-prone to render the match into English, or even to list enough examples to illuminate interesting cases. Maybe there is some external spec we can refer to?

:+1:, this looks good. Probably should have been in the original version; sorry!

Note that the regex itself is almost verbatim from [PEP 440 Appendix B](https://www.python.org/dev/peps/pep-0440/#appendix-b-parsing-version-strings-with-regular-expressions). I split it into multiple parts for line length and adapted/simplified the additional bit to match the local version segment, which for some reason is missing from the sample `is_canonical`.